casper/src/main/scala/io/casperlabs/casper/deploybuffer/DeployBuffer.scala

```
@@ -0,0 +1,276 @@
+  // Deploys that have been processed at least once,
+  // waiting to be finalized or orphaned
+  private val ProcessedStatusCode = 1
+  // Deploys that have been discarded for some reason and should be deleted after a while
+  private val DiscardedStatusCode = 2
+
+  import DeployBufferImpl.metricsSource
+
+  private implicit val metaByteString: Meta[ByteString] =
+    Meta[Array[Byte]].imap(ByteString.copyFrom)(_.toByteArray)
+  private implicit val readDeploy: Read[Deploy] =
+    Read[Array[Byte]].map(Deploy.parseFrom)
+
+  override def addAsPending(deploys: List[Deploy]): F[Unit] =
+    insertNewDeploys(deploys, PendingStatusCode)
+
+  override def addAsProcessed(deploys: List[Deploy]): F[Unit] =
+    insertNewDeploys(deploys, ProcessedStatusCode)
+
+  private def insertNewDeploys(
+      deploys: List[Deploy],
+      status: Int
+  ): F[Unit] = {
+    val writeToDeploysTable = Update[(ByteString, ByteString, ByteString)](
+      "INSERT INTO deploys (hash, account, data) VALUES (?, ?, ?)"
+    ).updateMany(deploys.map { d =>
+      (d.deployHash, d.getHeader.accountPublicKey, d.toByteString)
+    })
+
+    def writeToBufferedDeploysTable(currentTimeEpochSeconds: Long) =
+      Update[(ByteString, Int, ByteString, Long, Long)](
+        "INSERT INTO buffered_deploys (hash, status, account, last_accessed_at_epoch_millis, received_at) VALUES (?, ?, ?, ?, ?)"
+      ).updateMany(deploys.map { d =>
+          (
+            d.deployHash,
+            status,
+            d.getHeader.accountPublicKey,
+            currentTimeEpochSeconds,
+            currentTimeEpochSeconds
+          )
+        })
+        .void
+
+    for {
+      t <- Time[F].currentMillis
+      _ <- (writeToDeploysTable >> writeToBufferedDeploysTable(t))
+            .handleError {
+              case ex: SQLiteException
+                  if ex.getResultCode == SQLiteErrorCode.SQLITE_CONSTRAINT_PRIMARYKEY =>
+            }
+            .transact(xa)
+      _ <- updateMetrics()
```

What happens when you try to insert 5 new deploys and 1 of them is not new? Do the other 4 get inserted, or does it abort the whole batch due to the primary key violation?

---

Good catch! Previously it was silently failing a batch. I've changed so it will try to add as much as it can ignoring duplicates. Updated the test as well.

---

Cool, so you don't need the error handling clause any more. Unless you want to keep it for reference?

---

You were right, I can safely remove this check.